core/src/main/web/app/helpers/helper.js

```diff
@@ -63,7 +63,112 @@
       newSession: function(empty) {
         return bkCoreManager.newSession(empty);
       },
-
+      getBaseUrl: function() {
+        return bkUtils.getBaseUrl();
+      },
+      openWindow: function(path) {
+        bkUtils.openWindow(path);
+      },
+      openStaticWindow: function(path) {
+        bkUtils.openStaticWindow(path);
+      },
+      openBrowserWindow: function(path) {
+        bkUtils.openBrowserWindow(path);
+      },
+      // Save file with electron or web dialog
+      saveWithDialog: function(thenable) {
+        if (bkUtils.isElectron){
+          var BrowserWindow = bkUtils.Electron.BrowserWindow;
+          var Dialog = bkUtils.Electron.Dialog;
+          var thisWindow = BrowserWindow.getFocusedWindow();
+          var path = showElectronSaveDialog(thisWindow, options).then(function(path) {
+            if (path === undefined){
+              saveFailed('cancelled');
+              return;
+            }
+            bkUtils.httpPost('rest/file-io/setWorkingDirectory', { dir: path });
+            var ret = {
+              uri: path,
+              uriType: 'file'
+            };
+            bkSessionManager.dumpDisplayStatus();
+            var saveData = bkSessionManager.getSaveData();
+            var fileSaver = bkCoreManager.getFileSaver(ret.uriType);
+            var content = saveData.notebookModelAsString;
+            fileSaver.save(ret.uri, content, true).then(function() {
+              thenable.resolve(ret);
+            }, function(reason) {
+              thenable.reject(reason);
+            });
+          });
+          return thenable.promise.then(saveDone, saveFailed);
+        } else {
+          thenable = savePromptChooseUri();
+          return thenable.then(saveDone, saveFailed);
+        }
+      },
+      showElectronSaveDialog: function() {
+        var BrowserWindow = bkUtils.Electron.BrowserWindow;
+        var Dialog = bkUtils.Electron.Dialog;
+        var def = bkUtils.newDeferred();
+        bkUtils.getWorkingDirectory().then(function(defaultPath) {
+          var options = {
+            title: 'Save Beaker Notebook',
+            defaultPath: defaultPath,
+            filters: [
+              { name: 'Beaker Notebook Files', extensions: ['bkr'] }
+            ]
+          };
+          var path = Dialog.showSaveDialog(options);
+          def.resolve(path);
+        });
+        return def.promise;
+      },
+      // Open file with electron or web dialog
+      openWithDialog: function(ext, uriType, readOnly, format) {
+        var promise;
```

Comment:
same as before, why not just return the promises directly as compared to a proxy promise object?